#  Assignment 5

This assignment covers material on matrices, norms, condition numbers and eigensystems. The assignment is due by 5pm on Friday the 17th of April.

The breakdown of the marks is as follows:

Question 0 - [1 mark]  
Question 1 - [7 marks]  
Questino 2 - [6 marks]  
Question 3 - [7 marks]  
Question 4 - [9 marks]  

# Question 0

You should have saved this notebook to your ACM20030-Assignments folder. Now:
- commit the notebook to your local repository
- push the changes to the GitHub

In [1]:
import numpy as np
from numpy import linalg as la
np.set_printoptions(linewidth=200)  # This stops line wrapping of larger matrices

# Question 1 - practice manipulating arrays

This questions asks you write a few functions for manipulating arrays of data

## Q1 a

Write a function that checks if a given $n\times n$ matrix is symmetric. Use only loops and `.shape` function from numpy (do not use any other special numpy functions).

In [31]:
def checkSymmetric(a):
    n=a.shape[0]
    if(not n == a.shape[1]):
        return "matrix is not square"
    
    #checks if the upper right triangle (not including the centre line of entries) equals the bottom left triangle
    for i in range(0,n):
        for j in range(i+1,n):
            if(not a[i,j]==a[j,i]):
                return False
                
    return True

Test your function on the two matrices below

In [28]:
A = np.random.rand(5,5)
SymMatrix = (A + A.T)/2

In [33]:
checkSymmetric(A)

False

In [32]:
checkSymmetric(SymMatrix)

True

## Q1 b

Write a function to transpose a general $n\times n$ matrix. Do this by writing loops and not using any special functions like `np.transpose()`, `A.T`, or any functions from any other library.

In [37]:
def matrixTranspose(a):
    n=a.shape[0]
    if(not n == a.shape[1]):
        return "matrix is not square"
    out = np.zeros((n,n))
    for i in range(n):
        for j in range(n):
            out[j,i]=a[i,j]
    #print(out)
    return out

def matrixEquals(a,b):
    if (not len(a)==len(b)):
        return False
    if (not len(a[0])==len(b[0])):
        return False
    for i in range(len(a)):
        for j in range(len(a[0])):
            if(not a[i,j]==b[i,j]):
                return False
    return True

Check your code by transposing the below random matrix and checking that $A^T \neq A$ and $(A^T)^T - A = 0$

In [38]:
A = np.random.rand(6,6)
At = matrixTranspose(A)
print("A:",A,"\n")
print("A^T:",At,"\n")
print("A==A^T?",matrixEquals(A,At),"\n")
print("(A^T)^T-A==0?",matrixEquals(matrixTranspose(At)-A,np.zeros((len(A),len(A)))),"\n")


A: [[5.74551137e-01 7.75607173e-01 2.23369205e-01 2.44504450e-01 1.52875102e-01 8.69987076e-01]
 [1.41638130e-01 9.21283813e-01 6.65275942e-02 1.21731992e-01 6.23843651e-01 5.22042072e-01]
 [5.46843684e-01 9.79304031e-02 1.89389775e-01 9.60082145e-01 2.54474952e-01 5.22642868e-01]
 [4.79476369e-01 2.85232690e-01 2.66043006e-01 6.24718002e-01 8.79129067e-01 1.31779453e-01]
 [4.90666409e-01 1.68618141e-01 5.45699285e-01 2.88803201e-01 3.09383494e-01 3.89282511e-01]
 [2.40898749e-04 1.27377758e-01 8.98408243e-01 9.32229904e-01 2.34516444e-01 3.10292011e-01]] 

A^T: [[5.74551137e-01 1.41638130e-01 5.46843684e-01 4.79476369e-01 4.90666409e-01 2.40898749e-04]
 [7.75607173e-01 9.21283813e-01 9.79304031e-02 2.85232690e-01 1.68618141e-01 1.27377758e-01]
 [2.23369205e-01 6.65275942e-02 1.89389775e-01 2.66043006e-01 5.45699285e-01 8.98408243e-01]
 [2.44504450e-01 1.21731992e-01 9.60082145e-01 6.24718002e-01 2.88803201e-01 9.32229904e-01]
 [1.52875102e-01 6.23843651e-01 2.54474952e-01 8.79129067e-

## Q1c

Write a function that finds the maximum entry in a matrix (where the matrix does not need to be square). Do this by writing loops and not using any special functions like `np.max` or any functions from any other library.

In [45]:
def matrixMax(a):
    
    maxIndex = (0,0)
    maxValue = a[maxIndex]
    #print(maxValue)
    for i in range(len(a)):
        for j in range(len(a[0])):
            if(a[i,j]>maxValue):
                maxValue = a[i,j]
                maxIndex = (i,j)
    return maxValue

Check your result by comparing to `np.max(A)` for the below random matrix

In [47]:
A = np.random.rand(4,5)
#print(A[0,0])
print("My maximum value: %f\nNumPy's maximum value: %f"%(matrixMax(A),np.max(A)))

My maximum value: 0.804612
NumPy's maximum value: 0.804612


# Question 2:  matrix norms

In the lectures we focused on the $L^2$-norm. We also briefly mentioned the $L^1$ and $L^\infty$ norm. Let's look at calculating those as well for a general $m\times n$ matrix.

# Q2a

The $L^1$ norm is given by

$$ \lVert A \rVert_1 = \max_j \sum^m_{i=1} |a_{ij}| $$

i.e., it is the max of the sums of the absolute values of each column. Write a function that computes the $L^1$-norm using only loops. The function should use no special numpy functions except `np.abs()`.

In [49]:
def L1Norm(a):
    maxValue = 0
    maxbool=True
    for j in range(len(a[0])):
        currentSum = 0
        for i in range(len(a)):
            currentSum +=np.abs(a[i,j])
        if(maxbool):
            maxbool=False
            maxValue=currentSum
        else:
            if currentSum>maxValue:
                maxValue = currentSum
    return maxValue

Check your result against NumPy's `la.norm(A,1)` function for the following random matrix

In [50]:
A = np.random.rand(3,4)
print("My value for the L^1-norm:",L1Norm(A))
print("NumPy's L^1-norm:",la.norm(A,1))

My value for the L^1-norm: 2.4965747882569773
NumPy's L^1-norm: 2.4965747882569773


## Q2b

The $L^\infty$ norm is given by 

$$\lVert A \rVert_\infty = \max_i \sum_{j=1}^n |a_{ij}|$$

i.e., it is the max of the sums of the absolute values of each row. Write a function that computes the $L^\infty$-norm using only loops and no special numpy functions.

In [51]:
def LInfNorm(a):
    maxValue = 0
    maxbool=True
    for i in range(len(a)):
        currentSum = 0
        for j in range(len(a[0])):
            currentSum +=np.abs(a[i,j])
        if(maxbool):
            maxbool=False
            maxValue=currentSum
        else:
            if currentSum>maxValue:
                maxValue = currentSum
    return maxValue

Check your result against NumPy's `la.norm(A,np.inf)` function for the following random matrix

In [52]:
A = np.random.rand(3,4)
print("My value for the L^inf-norm:",LInfNorm(A))
print("NumPy's L^inf-norm:",la.norm(A,np.inf))

My value for the L^inf-norm: 2.7402473435424035
NumPy's L^inf-norm: 2.7402473435424035


# Question 3: ill-conditions systems of equations

In this question you can use any of NumPy's functions.

## Q3a

Which of the following three matrices, $A1, A2, A3$, are ill-conditioned. State the reason for your conclusion.

In [53]:
A1 = np.array([[0.988746460106, 0.351486109477, 0.024911477532, 0.384039749683],
       [0.282104657306, 0.239884151775, 0.36170151991 , 0.903082266861],
       [0.59333290915 , 0.637178863968, 0.241229214635, 0.685595037901],
       [0.013591747828, 0.917926970364, 0.458689093885, 0.157821121254]])

A2 = np.array([[ 2.852416473483e-01,  3.803483118694e-01,  2.697704226632e-01,  6.810611849029e-01],
       [ 6.622827828376e-01,  6.447943415836e-01,  4.387842058829e-01,  1.530575702376e+00],
       [ 3.672130860703e-01,  8.309522079704e-01,  9.975487474007e-01,  5.678296337101e-01],
       [ 4.036490000599e-01,  6.865978225967e-06,  8.788095913773e-01, -7.150472427928e-02]])

A3 = np.array([[0.131923621844, 0.245316154195, 0.732362341613, 0.957047891764],
       [0.91290446184 , 0.512312007959, 0.391938831444, 0.092509684772],
       [0.648708523652, 0.608576675797, 0.576773587276, 0.065385122929],
       [0.502955480393, 0.084082960221, 0.659487390444, 0.662958739412]])

In [54]:
print("L^2 condition number of A1",la.cond(A1,2))
print("L^2 condition number of A2",la.cond(A2,2))
print("L^2 condition number of A3",la.cond(A3,2))


L^2 condition number of A1 21.55246648838184
L^2 condition number of A2 13334945724.122156
L^2 condition number of A3 15.122449699410616


A2 has a large condition number ($\simeq 13334945724$), making it ill-conditioned.
A1 and A3 have low condition numbers, so they are not ill-conditioned

# Q3b

With $b = [1,2,3,4]$ and $\delta b = [0.1,0,0,0]$ for $A2$ and $A3$ calculate the percentage change in $x$ due to a small change in $b$, i.e., calculate 

$$100\times\frac{\lVert \delta x_i\rVert_2}{\lVert x_i \rVert_2}\quad\text{and}\quad 100\times\frac{\lVert \delta b\rVert_2}{\lVert b \rVert_2}$$

where $(x_i+\delta x_i)  = A_i^{-1}(b+\delta b)$ and $i\in\{2,3\}$.

In [56]:
b=np.array([1,2,3,4])
db=np.array([.1,0,0,0])

#la.norm(db,2)/la.norm(b,2)
b_per = 100*la.norm(db,2)/la.norm(b,2)
#la.norm(,2)/la.norm(b,2)

x2 = np.dot(la.inv(A2),(b))
x2_dash = np.dot(la.inv(A2),(b+db))
dx2 = x2_dash-x2
x2_per = 100*la.norm(dx2,2)/la.norm(x2,2)

x3 = np.dot(la.inv(A3),(b))
x3_dash = np.dot(la.inv(A3),(b+db))
dx3 = x3_dash-x3
x3_per = 100*la.norm(dx3,2)/la.norm(x3,2)


#np.transpose(A2)*(b+db)

print("change in b(%%): %0.3f%%" % b_per)
print("(for maxtrix A2) change in x_2(%%):%0.3f%%" % x2_per)
print("(for maxtrix A3) change in x_3(%%):%0.3f%%" % x3_per)

change in b(%): 1.826%
(for maxtrix A2) change in x_2(%):512.621%
(for maxtrix A3) change in x_3(%):2.174%


Which system of equations is more sensitive to small changes in $b$? How could you have guessed this from the answer to the Q3a?

# Answer here

A2 is more sensitive to changes in b.

We could have guessed this from A2's large condition number

# Question 4: eigensystems

In this question we look at finding eigenvalues and eigenvectors

## Q4a

You may use any of NumPy's functions for this sub-question.

Find the eigenvalues and eigenvectors for the below matrix. Print out the eigenvalues sorted by descending value and explicitly show that eigenvalues/eigenvector pairs satisfy $Ax = \lambda x$.

In [58]:
A = np.array([[1,4,5],[4,5,4],[6,7,9]])

In [70]:
eigs = la.eig(A)

#code to sort the indexes of the eigenvalues in descending order
values = eigs[0]
vectors = eigs[1]    
    
order = []
remove = [0,1,2]
k = len(remove)
for i in range(k):
    largestIn = remove[0]
    largestVal = values[remove[0]]
    for j in remove:
        if values[j] > largestVal:
            largestVal = values[j]
            largestIn = j
    remove.remove(largestIn)
    order.append(largestIn)
    
c=1
for i in order:
    print("eigenvalue %d:"%(c),values[i],"\neigenvector %d:"%(c),vectors[i],"\n")
    c+=1
c=1
for i in order:
    print("Ax_%d:"%(c),np.dot(A,vectors[:,i]),", \nlam_%d x_%d:"%(c,c),np.dot(values[i],vectors[:,i]),)
    diff = (np.dot(A,vectors[:,i])-np.dot(values[i],vectors[:,i]))
    print("Ax_%d-lam_%dx_%d:[%0.5f,%0.5f,%0.5f]"%(c,c,c,diff[0],diff[1],diff[2]),"\n")
    c+=1

eigenvalue 1: 15.785693179172181 
eigenvector 1: [-0.39252008 -0.89589878  0.09818766] 

eigenvalue 2: 1.2659532728352207 
eigenvector 2: [-0.80534308  0.25675168  0.6248558 ] 

eigenvalue 3: -2.0516464520074047 
eigenvector 3: [-0.44424151  0.36255198 -0.77454142] 

Ax_1: [ -6.1962015   -7.01266016 -12.71289871] , 
lam_1 x_1: [ -6.1962015   -7.01266016 -12.71289871]
Ax_1-lam_1x_1:[-0.00000,-0.00000,0.00000] 

Ax_2: [ 0.12430099 -0.98053324  0.79103825] , 
lam_2 x_2: [ 0.12430099 -0.98053324  0.79103825]
Ax_2-lam_2x_2:[-0.00000,-0.00000,-0.00000] 

Ax_3: [ 1.83806756 -0.74382849 -0.52676368] , 
lam_3 x_3: [ 1.83806756 -0.74382849 -0.52676368]
Ax_3-lam_3x_3:[0.00000,0.00000,-0.00000] 



# Q4b

Without using NumPy's `la.eig()` (or similar functions), find the 4 eigenvalues and eigenvectors of the following matrix. Hint: Use the power method for the max eigenvalue (given below in `MaxEigenvalue`), the power method applied to the inverse for the smallest eigenvalue, and apply Hotelling's deflation twice to find the second and third eigenvalues/eigenvectors.

In [71]:
A = np.array([[1.935052353404, 0.887037260957, 1.534235483277, 1.370691224125],
       [0.887037260957, 0.81275989158 , 0.610238622918, 0.999150521052],
       [1.534235483277, 0.610238622918, 1.562260333538, 1.007843062517],
       [1.370691224125, 0.999150521052, 1.007843062517, 1.813258814483]])

In [72]:
def MaxEigenvalue(A, err):
    (m,n) = A.shape
    if(m != n):
        print("Matrix must be square")
        return
    x = np.random.rand(m)
    
    lam = 0.1
    lamprev = 1
    while np.abs(1-lam/lamprev) > err:
        p = np.dot(A,x)
        lamprev = lam
        lam = np.max(p)
        x = p/lam

    x = x/la.norm(lam)
    return (lam,x)

In [73]:
eigenList = [0,0,0,0]

#maximum eigenvalue
eigenList[0] = MaxEigenvalue(A,10**-12)

#normalise
sum = 0
for i in eigenList[0][1]:
    sum+= i**2
eigenList[0] = (eigenList[0][0], np.array(eigenList[0][1]/np.sqrt(sum)))

#second to maximum eigenvalue
matrix_eig_1 = A - np.dot(eigenList[0][0],(np.outer((eigenList[0][1]),eigenList[0][1])))
eigenList[1] = MaxEigenvalue(matrix_eig_1,10**-12)

#normalise
sum = 0
for i in eigenList[1][1]:
    sum+= i**2
    
eigenList[1] = (eigenList[1][0], np.array(eigenList[1][1]/np.sqrt(sum)))


#minimum eigenvalue

# The power method seems to get stuck on A - lam_max * identity
# Instead I used (lam_max * identity - A)x_i = (lam_max - lam_i)x_i
matrix_eig_3 = np.identity(4)*eigenList[0][0] - A
eigenList[3] = MaxEigenvalue(matrix_eig_3,10**-12)

#normalise
sum = 0
for i in eigenList[3][1]:
    sum+= i**2
    
eigenList[3] = (eigenList[0][0] - eigenList[3][0], np.array(eigenList[3][1]/np.sqrt(sum)))


#second to minimum eigenvalue
matrix_eig_2 = matrix_eig_3 - np.dot(eigenList[3][0],(np.outer((eigenList[3][1]),eigenList[3][1])))
eigenList[2] = MaxEigenvalue(matrix_eig_2,10**-12)

#normalise
sum = 0
for i in eigenList[2][1]:
    sum+= i**2
    
eigenList[2] = (eigenList[0][0] - eigenList[2][0], np.array(eigenList[2][1]/np.sqrt(sum)))

for i in range(len(eigenList)):
    print("Eigenvalue: %f \nEigenvector:" %(eigenList[i][0]),eigenList[i][1],"\n")


Eigenvalue: 4.914804 
Eigenvector: [0.59947641 0.33318806 0.49528112 0.53320761] 

Eigenvalue: 0.850582 
Eigenvector: [-0.30724064  0.3894447  -0.58292464  0.64353313] 

Eigenvalue: 0.215128 
Eigenvector: [-0.38116859 -0.67389895  0.34000478  0.53382343] 

Eigenvalue: 0.142817 
Eigenvector: [-0.63319959  0.53214542  0.54707607 -0.12879173] 



Check your answers against Numpy's results from `la.eig(A)`

In [74]:
eigs = la.eig(A)

#code to sort the indexes of (NumPy's) eigenvalues in descending order
values = eigs[0]
vectors = eigs[1]
    
order = []
remove = [0,1,2,3]
k = len(remove)
for i in range(k):
    largestIn = remove[0]
    largestVal = values[remove[0]]
    for j in remove:
        if values[j] > largestVal:
            largestVal = values[j]
            largestIn = j
    remove.remove(largestIn)
    order.append(largestIn)       

    
c=0 
for i in order:
    print("My Eigenvalue: %f\nNumPy's Eigenvalue: %f," %(eigenList[c][0],values[i]))
    print("My Eigenvector:",eigenList[c][1],"\nNumPy's Eigenvector",vectors[:,i],"\n")
    c+=1


My Eigenvalue: 4.914804
NumPy's Eigenvalue: 4.914804,
My Eigenvector: [0.59947641 0.33318806 0.49528112 0.53320761] 
NumPy's Eigenvector [0.59947641 0.33318806 0.49528112 0.53320761] 

My Eigenvalue: 0.850582
NumPy's Eigenvalue: 0.850582,
My Eigenvector: [-0.30724064  0.3894447  -0.58292464  0.64353313] 
NumPy's Eigenvector [ 0.30724064 -0.3894447   0.58292464 -0.64353313] 

My Eigenvalue: 0.215128
NumPy's Eigenvalue: 0.215128,
My Eigenvector: [-0.38116859 -0.67389895  0.34000478  0.53382343] 
NumPy's Eigenvector [-0.3811686  -0.67389895  0.34000478  0.53382343] 

My Eigenvalue: 0.142817
NumPy's Eigenvalue: 0.142817,
My Eigenvector: [-0.63319959  0.53214542  0.54707607 -0.12879173] 
NumPy's Eigenvector [-0.63319959  0.53214542  0.54707607 -0.12879172] 

